File: core/src/com/unciv/models/stats/FullStats.java
package com.unciv.models.stats;

import java.util.HashMap;

public class FullStats extends CivStats // also used for hex stats, since it's basically the same
{
    public float production = 0;
    public float food = 0;

    public FullStats() {
    }

    public FullStats(FullStats other){
        add(other);
    }

    public void add(FullStats other){
        gold +=other.gold;
        science +=other.science;
        happiness +=other.happiness;
        culture +=other.culture;
        food +=other.food;
        production +=other.production;
    }

    public FullStats minus(){
        FullStats sub = new FullStats();
        sub.gold=-gold;
        sub.science=-science;
        sub.happiness=-happiness;
        sub.culture=-culture;
        sub.food=-food;
        sub.production=-production;
        return sub;
    }

    public FullStats multiply(float number){
        FullStats mul = new FullStats();
        mul.gold=gold*number;
        mul.science=science*number;
        mul.happiness=happiness*number;
        mul.culture=culture*number;
        mul.food=food*number;
        mul.production=production*number;
        return mul;
    }

    public String display(float value, String name){
        return ", " + (value>0 ? "+" : "") + Math.round(value) + " "+name;
    }

    public String toString() {
        StringBuilder valuableParts = new StringBuilder();
        if (production != 0) valuableParts.append(display(production,"Production"));
        if (food != 0) valuableParts.append(display(food,"Food"));
        if (gold != 0) valuableParts.append(display(gold,"Gold"));
        if (science != 0) valuableParts.append(display(science,"Science"));
        if (happiness != 0) valuableParts.append(display(happiness,"Happiness"));
        if (culture != 0) valuableParts.append(display(culture,"Culture"));
        if (valuableParts.length() == 0) return "";
        valuableParts.delete(0,1);
        return valuableParts.toString();
    }

    public HashMap<String,Integer> toDict(){
        HashMap<String,Integer> dict = new HashMap<String, Integer>();

        dict.put("Production", (int) production);
        dict.put("Food", (int) food);
        dict.put("Gold", (int) gold);
        dict.put("Science", (int) science);
        dict.put("Culture", (int) culture);
        return dict;
    }

}

File: core/src/com/unciv/models/stats/FullStats.kt
package com.unciv.models.stats

import java.util.HashMap


open class FullStats : CivStats()
{
    @JvmField var production: Float = 0f
    @JvmField var food: Float = 0f

    fun add(other: FullStats) {
        production += other.production
        food += other.food
        super.add(other)
    }

    fun clone():FullStats {
        val FS = FullStats()
        FS.add(this)
        return FS
    }

    fun getMinus(): FullStats {
        val stats = FullStats()
        stats.food = -food
        stats.food = -food
        stats.gold = -gold
        stats.science = -science
        stats.culture = -culture
        stats.happiness = -happiness
        return stats;
    }

    operator fun times(number: Float): FullStats{
        val FS = FullStats()
        FS.production = production * number
        FS.food = food*number
        FS.gold = gold*number
        FS.science = science*number
        FS.culture = culture*number
        FS.happiness = happiness*number
        return FS
    }

    override fun toString(): String {
        return toDict().filter{it.value!=0}.map{it.key+": "+it.value}.joinToString(",")
    }

    fun toDict(): HashMap<String, Int> {
        return hashMapOf("Production" to production.toInt(),
                "Food" to food.toInt(),
                "Gold" to gold.toInt(),
                "Science" to science.toInt(),
                "Culture" to culture.toInt()
        )
    }

}
